cfme/common/provider.py

```diff
@@ -728,3 +831,65 @@ def cleanup_vm(vm_name, provider):
     except:
         # The mgmt_sys classes raise Exception :\
         logger.warning('Failed to clean up VM %s on provider %s', vm_name, provider.key)
+
+
+class DefaultEndpoint(object):
+    credential_class = Credential
+    name = 'default'
+
+    def __init__(self, **kwargs):
+        for key, val in kwargs.items():
+            if key == 'credentials' and not isinstance(val, (Credential, TokenCredential)):
+                val = self.credential_class.from_config(val)
+            setattr(self, key, val)
+
+        if not hasattr(self, 'credentials'):
+            logger.warn("credentials weren't passed "
+                        "for endpoint {}".format(self.__class__.__name__))
+
+    @property
+    def view_value_mapping(self):
+        return {'hostname': self.hostname}
+
+
+class CANDUEndpoint(DefaultEndpoint):
+    credential_class = CANDUCredential
+    name = 'candu'
+
+    @property
+    def view_value_mapping(self):
+        return {'hostname': self.hostname,
+                'api_port': self.api_port,
+                'database_name': self.database}
+
+
+class EventsEndpoint(DefaultEndpoint):
+    credential_class = EventsCredential
+    name = 'events'
+
+    @property
+    def view_value_mapping(self):
+        return {'event_stream': self.event_stream,
+                'security_protocol': self.security_protocol,
+                'hostname': self.hostname,
+                'api_port': self.api_port,
+                }
+
+
+class SSHEndpoint(DefaultEndpoint):
+    credential_class = SSHCredential
```

Comment:
I have created endpoint forms as multi tab views repeating real form behaviour and appearance. 
So, each endpoint class is only one tab in such form.
I can split such form to many small ones but I'll have to construct such form in create/update methods them and I suspect it will be even less obvious.